### PR TITLE
Add game status to game

### PR DESCRIPTION
**Pull request by cameron-prc**
The game is now playable with a responsive front end.
It does not currently prevent duplicate guesses

### CODE REVIEW

**Comment by trezona-lecomte** [comment]
I recommend having a read of https://chris.beams.io/posts/git-commit/ - I will create a confluence page with our own guidelines at some point but for now this should serve as a good starting point.

**Comment by trezona-lecomte** [comment]
It would be good to define a `root` route

File: app/helpers/guess_helper.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module GuessHelper
+  def generate_obfuscated_word(word, guessed_letters)
+    string = ""
+
+    word.split("").each do |char|
+      string += guessed_letters.include?(char) ? " #{char} " : " _ "
+    end
+
+    string
```
Comment (trezona-lecomte) [inline app/helpers/guess_helper.rb]: This method can be simplified to a one-liner using http://ruby-doc.org/core-2.6.2/Enumerable.html#method-i-reduce, I'll leave it up to you to implement.

File: app/helpers/guess_helper.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module GuessHelper
+  def generate_obfuscated_word(word, guessed_letters)
+    string = ""
+
+    word.split("").each do |char|
```
Comment (trezona-lecomte) [inline app/helpers/guess_helper.rb]: This is the `guessed_letters` concept used elsewhere inside `Game` - would be good to consolidate so there is only one place in the code that knows how to get the target letters of a game.

File: app/helpers/guess_helper.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+module GuessHelper
+  def generate_obfuscated_word(word, guessed_letters)
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/helpers/guess_helper.rb]: Added an rspec covering all common paths in 04186d9

File: app/models/game.rb [hunk 4]
```diff
@@ -4,4 +4,31 @@ class Game < ApplicationRecord
   validates :target_word, presence: true
   validates :lives, presence: true
   validates :lives, numericality: { only_integer: true, greater_than: 0 }
+
+  def won?
+    (target_word.split("") - guessed_letters).empty?
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/game.rb]: Extracted in 72df69f and corrected from split into chars in 2862866 @trezona-lecomte

File: app/models/game.rb [hunk 5]
```diff
@@ -4,4 +4,31 @@ class Game < ApplicationRecord
...
+
+  def lost?
+    incorrect_guesses >= lives
+  end
+
+  def active?
+    !(lost? or won?)
+  end
+
+  def incorrect_guesses
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/game.rb]: This has been updated in 57df561 @trezona-lecomte

File: app/models/game.rb [hunk 6]
```diff
@@ -4,4 +4,31 @@ class Game < ApplicationRecord
...
+    valid_letters = target_word.split("")
+    incorrect_guesses = 0
+
+    guesses.where.not(id: nil).each do |guess|
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/game.rb]: This is covered in 57df561 with an explanation as to why it is required @trezona-lecomte

File: spec/helpers/guess_helper_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
+require "spec_helper"
+
+RSpec.describe GuessHelper, type: :helper do
+  let(:target_word) { "word" }
+  let(:guessed_letters) { [] }
+
+  describe "generate_obfuscated_word" do
+    context "with no guesses" do
+      it "returns a string of padded underscores only" do
+        helper.generate_obfuscated_word(target_word, guessed_letters).should == " _  _  _  _ "
```
Comment (trezona-lecomte) [inline spec/helpers/guess_helper_spec.rb]: `should ==` is old syntax, we should use `expect`:

```suggestion
        expect(helper.generate_obfuscated_word(target_word, guessed_letters)).to eq(" _  _  _  _ ")
```

File: app/helpers/guess_helper.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (trezona-lecomte) [reply app/helpers/guess_helper.rb]: 💯

File: app/helpers/guess_helper.rb (hunk 2 above)
Comment (cameron-prc) [reply app/helpers/guess_helper.rb]: I would argue that the logic is different as target_letters used in the game model is a array of unique letters used to determine is each letter has been guessed whereas this is an array of all letters in order used to determine how to display the word and is not used elsewhere.

File: app/helpers/guess_helper.rb (hunk 1 above)
Comment (cameron-prc) [reply app/helpers/guess_helper.rb]: Addressed in f03e327 @trezona-lecomte

File: app/models/game.rb [hunk 8]
```diff
@@ -4,4 +4,31 @@ class Game < ApplicationRecord
...
+      incorrect_guesses += 1 unless valid_letters.include? guess.letter
+    end
+
+    incorrect_guesses
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/game.rb]: Addressed in bd26236 @trezona-lecomte

File: app/models/game.rb [hunk 9]
```diff
@@ -4,4 +4,31 @@ class Game < ApplicationRecord
...
+  def won?
+    (target_word.split("") - guessed_letters).empty?
+  end
+
+  def lost?
+    incorrect_guesses >= lives
+  end
+
+  def active?
+    !(lost? or won?)
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/game.rb]: Addressed in 4f2705a @trezona-lecomte

File: spec/helpers/guess_helper_spec.rb (hunk 7 above)
Comment (cameron-prc) [reply spec/helpers/guess_helper_spec.rb]: Addressed in 2877447 @trezona-lecomte

File: app/views/games/show.html.erb [hunk 10]
```diff
@@ -1,3 +1,16 @@
 <p id="notice"><%= notice %></p>
 
+<% unless @game.active? %>
+    <p>Game <%= @game.lost? ? "lost" : "won" %></p>
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/views/games/show.html.erb]: Addressed in 094900f @trezona-lecomte

File: app/views/games/index.html.erb [hunk 11]
```diff
@@ -12,7 +12,8 @@
   <tbody>
     <% @games.each do |game| %>
       <tr>
-        <td><%= link_to 'Show', game %></td>
+        <td>Game: <%= game.id %></td>
+        <td><%= link_to "#{ game.active? ? "Continue" : (game.lost? ? "Lost" : "Won") }", game %></td>
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/views/games/index.html.erb]: Addressed in dbb046f @trezona-lecomte

File: app/models/guess.rb [hunk 12]
```diff
@@ -3,4 +3,13 @@ class Guess < ApplicationRecord
...
   validates :letter, presence: true
   validates :letter, length: { is: 1 }
+  validate :validates_game_is_active
+
+  private
+
+  def validates_game_is_active
+    return unless game
+
+    errors.add(:game, "The game has already been completed") unless game.active?
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/models/guess.rb]: Addressed in 2695f15 @trezona-lecomte

File: app/views/games/show.html.erb [hunk 13]
```diff
@@ -1,3 +1,16 @@
...
+<% end %>
+<p><%= generate_obfuscated_word(@game.target_word, @game.guessed_letters) %></p>
+
+<p>Lives remaining: <%= @game.lives - @game.incorrect_guesses %></p>
```
[the remark this answers is not in the document]
Comment (cameron-prc) [reply app/views/games/show.html.erb]: Addressed in 094900f @trezona-lecomte

File: app/models/game.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (trezona-lecomte) [reply app/models/game.rb]: Nice, good explanation in the commit message.

This does have the side effect of re-querying the `guesses` table repeatedly, so perhaps we could accomplish the same while still making use of the associations cache:

```diff
-guesses.where.not(id: nil)
+guesses.select(&:persisted?)
```

File: app/models/game.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (trezona-lecomte) [reply app/models/game.rb]: 👍

File: app/helpers/guess_helper.rb (hunk 2 above)
Comment (trezona-lecomte) [reply app/helpers/guess_helper.rb]: Cool - makes sense

**Comment by trezona-lecomte** [comment]
Looks good! One more suggestion regarding the `guesses` query, otherwise good-to-go 🎉

**Pull request merged**